xdsl/dialects/riscv.py

```diff
@@ -2990,6 +2991,15 @@ class FSwOp(RsRsImmFloatOperation):
 
 # endregion
 
+
+def _parse_int_or_str_literal(parser: Parser, msg: str) -> int | str:
+    if (term := parser.parse_optional_integer()) is not None:
+        return term
+    if (term := parser.parse_optional_str_literal()) is not None:
+        return term
+    parser.raise_error(msg)
```

Comment:
I would recommend something like this:

```suggestion
def _parse_optional_immediate_value(parser: Parser, integer_type: type[IntegerType]) -> IntegerAttr | LabelAttr | None:
    """
    Parse an optional immediate value. If an int is parsed, an integer attr with the specified type is created.
    """
    if (term := parser.parse_optional_integer()) is not None:
        return IntegerAttr(immediate, type)
    if (term := parser.parse_optional_str_literal()) is not None:
        return LabelAttr(term)
    return None

def _parse_immediate_value(parser: Parser, integer_type: type[IntegerType]) -> IntegerAttr | LabelAttr:
    return parser.expect(lambda: _parse_optional_literal_value(parser, integer_type), "literal value")
```

It feels like the error message should be the same for all clients, and not a part of the helper API. We might have optional immediate values so an optional helper would also be nice.